### ISSUE TITLE
UnusedLocalVariable False-Positive when only second parameter is used

### ISSUE THREAD

**Issue by digilist** [phpmd/phpmd#278]
I am experiencing an error when I define a closure and only use the second argument.

For example, I experienced it with the Symfony Options Resolver:

```php
$resolver = new OptionsResolver();
$resolver->setRequired('startDate');
$resolver->setNormalizer('startDate', function ($options, $value) {
            if ($value === null) {
                return null;
            }

            return DateTime::createFromTimestamp(strtotime($value));
        });
```

PHPMD blames me for the unused local variable $options.
> Avoid unused local variables such as '$options'.

The problem here is, that the interface is predefined by the library. It is not possible to skip the parameter. I am not sure this is a bug by itself, because sometimes it might be an actual mistake in the code. A possible solution would be to ignore all parameters before the last used parameter on closures.

**Comment by zachborboa** [comment]
phpmd should allow prefixing the variable with `$unused_` and ignore when named as such.

`function ($options, $value)` would become `function ($unused_options, $value)`

**Comment by yoosefi** [comment]
Good idea, @zachborboa, but in the wild i've seen unused vars prefixed with `$aux_`, `$unused_`, `$ditch_`, `$ignore_`, and just `$_`.  Maybe there could be a tag or a way to configure what var prefixes are ignored.

**Comment by yoosefi** [comment]
How about suppressing this in general if some preceding variables are unused, but not tail ones?

**Comment by ravage84** [comment]
No this is no bug of PHPMD. It is likely a design problem of the library or simply an edge case, which the library did not account for. In your example code, it looks like the latter, rather.

Anyway, just because a succeeding parameter is used, does not make the first one used or ignorable Implementing such a rule would lead to a blind spot, so this gets a :-1: from me.

And I do not like the idea of having __five__ variable prefixes to suppress the violations. I'd rather have one. All of these prefixes could be part of perfectly fine variable names that should not suppress anything. Also I do no tlike the use of an underscore, because that would only mean it could violate some naming conventions. So, this gets a :-1: from me, too.

I would be OK, though, to have an ``exceptions`` property, like in other Rules. That would make it more consistent, too.
@digilist if you want to create a new PR with such a exceptions property, I'd would be :cool: .

Last but not least, there is still the possibility to use annotations to suppress some violations.

http://phpmd.org/documentation/suppress-warnings.html
https://github.com/phpmd/phpmd/blob/2.3.2/src/main/php/PHPMD/Node/ASTNode.php#L81-L85

The plus side of this is that it would be self-documenting within the code, instead of relying on some exception magic. Sure, it's not very beautiful to clutter your code with them, but if you have to clutter your code with them, you might have a different problem. :smiling_imp:

**Comment by yoosefi** [comment]
I wasn't suggesting having a prefix. I was suggesting not having prefixes, and simply ignoring unused parameters if latter parameters are used.

"It is likely a design problem of the library or simply an edge case."

Can you elaborate? PHP core functions specify callbacks with arbitrary parameter ordering. Are you suggesting we take this issue upstream simply because you won't ignore unused parameters preceding a used one? It's trivial to leave out an unused tail parameter. A leading one is the issue.

Signatures in PHP can't be polymorphic, so I don't see how this is an "edge case".

:-1: for having to suppress warnings for something that shouldn't be a warning.

**Comment by ravage84** [comment]
@yoosefi I'm sorry. I was not answering to you specifically but to you all three of you.
The example I referred to was @digilist 's [code](https://github.com/phpmd/phpmd/issues/278#issue-65272982).

So, no I was not talking about PHP core code and thus no I don't want you to take this upstream.

Your view on the current issue seems to be quite narrow, specifically on implementing callbacks based on PHP core functions. You might work a lot with such code, but userland code can look and work very diferently. Ignoring all preceding unused parameters, just because a succeeding one is used, would lead to ignoring many valid violations.

What about adding a configuration property to ignore such preceding parameters?
Similar to ``allow-unused-foreach-variables`` already in ``UnusedLocalVariable``.
http://phpmd.org/rules/unusedcode.html#unusedlocalvariable
How about ``allow-unused-preceeding-parameters``, which would defaultt to ``false``?

That way existing checks would not be altered but still people could ignore them on demand.

I'd like to hear @manuelpichler 's opinion first, though.

**Comment by yoosefi** [comment]
@ravage84 having the option available would be wonderful. :+1:

**Comment by nenadalm** [comment]
I just found this issue and I agree that the option `allow-unused-preceding-parameters` would be nice.

**Comment by yoosefi** [comment]
For anyone wondering, to shut the inspection up I just `unset()` the damn thing in the first line of the function.

**Comment by jannes-io** [comment]
Some 6 years later, with PHP adopting anonymous functions more closely, this is starting to become a serious issue. (Think of the new first class callable syntax, shorthand anon functions,...)

Personally I feel like it'd be a good idea to revisit this issue, as it's still the first one that pops up in Google without a solid answer or solution. @yoosefi 's solution with `unset` isn't possible in the shorthand syntax.

> No this is no bug of PHPMD

Agreed, it is a feature request, not a bug imo.

> Anyway, just because a succeeding parameter is used, does not make the first one used or ignorable Implementing such a rule would lead to a blind spot, so this gets a 👎 from me.

PHPStorm code analysis for example does consider a variable used in this case. Inspection tools do not report unused variables if any of the succeeding variables are used.

> It is likely a design problem of the library or simply an edge case, which the library did not account for. In your example code, it looks like the latter, rather.

Developers don't want code quality checks on the libraries we use, we can only control our own code. So to get a warning in our code because of the order of arguments that was picked by a library maintainer is ridicules.

> And I do not like the idea of having five variable prefixes to suppress the violations. I'd rather have one. All of these prefixes could be part of perfectly fine variable names that should not suppress anything. Also I do no tlike the use of an underscore, because that would only mean it could violate some naming conventions. So, this gets a 👎 from me, too.

Valid, however, the simple `$_` is already a convention that many are used to, especially those coming from a javascript environment.

One solution without adapting PHPMD is to catch all arguments using argument spread.
Example:

3rd party interface:
```php
    /**
     * Returns the first element of this collection that satisfies the predicate p.
     *
     * @param Closure $p The predicate.
     * @psalm-param Closure(TKey, T):bool $p
     *
     * @return mixed The first element respecting the predicate,
     *               null if no element respects the predicate.
     * @psalm-return T|null
     */
    public function findFirst(Closure $p);
```

usage without 1st arg:
```php
$firstMatchingType = $collection->findFirst(fn (...$args) => $args[1]->getType('cool_type'));
```

But the drawback of this is immense, as losing type hints also in most cases stops IDEs from being able to autocomplete.

> Ignoring all preceding unused parameters, just because a succeeding one is used, would lead to ignoring many valid violations.

Unfortunately, the only way to properly tackle this issue is to make users aware that they should order their arguments in order of importance. Which is an impossible task.

Personal preference goes out to PHPStorm's solution of considering arguments used if a succeeding argument is used, but only in closures/callables/anonymous functions, not in actual functions or methods.
